lib/galaxy/tools/error_reports/plugins/gitlab.py

```diff
@@ -0,0 +1,230 @@
+                    for issue in gl_project.issues.list():
+                        if issue.state != 'closed':
+                            log.info("GitLab error reporting - Repo issue: " + str(issue.iid))
+                            self.issue_cache[issue_cache_key][issue.title] = issue.iid
+
+                # Generate information for the tool
+                tool_kw = {'tool_id': unicodify(job.tool_id), 'tool_version': unicodify(job.tool_version)}
+                error_title = u"""Galaxy Job Error: {tool_id} v{tool_version}""".format(**tool_kw)
+
+                # We'll re-use the email error reporter's template since GitLab supports HTML
+                error_reporter = EmailErrorReporter(dataset.id, self.app)
+                error_reporter.create_report(job.get_user(), email=kwargs.get('email', None),
+                                             message=kwargs.get('message', None),
+                                             redact_user_details_in_bugreport=self.redact_user_details_in_bugreport)
+
+                # The HTML report
+                error_message = error_reporter.html_report
+
+                log.info(error_title in self.issue_cache[issue_cache_key])
+                if error_title not in self.issue_cache[issue_cache_key]:
+                    # Create a new issue.
+                    issue = gl_project.issues.create({
+                        'title': error_title,
+                        'description': error_message
+                    })
+                    self.issue_cache[issue_cache_key][error_title] = issue.iid
+                else:
+                    # Add a comment to an issue...
+                    gl_url = "/".join([
+                        self.gitlab_base_url,
+                        "api",
+                        "v4",
+                        "projects",
+                        gitlab_urlencodedpath,
+                        "issues",
+                        str(self.issue_cache[issue_cache_key][error_title]),
+                        "notes"
+                    ])
+                    self.gitlab.http_post(gl_url, post_data={'body': error_message})
+
+                return ('Submitted error report to GitLab. Your issue number is <a href="%s/%s/issues/%s" '
+                        'target="_blank">#%s</a>.' % (self.gitlab_base_url, gitlab_projecturl,
+                                                      self.issue_cache[issue_cache_key][error_title],
+                                                      self.issue_cache[issue_cache_key][error_title]), 'success')
+
+            except ImportError as e:
+                log.error("GitLab error reporting - Please ensure that requests, urllib and urlparse are available. "
+                          "Exception information: " + e.message)
+                return ('Error occured while submitting error report to GitLab. Report could not be submitted due to '
```

Comment:
I'd say that since these are the messages that bubble up to the user, you don't need to (and maybe shouldn't) expose that much debugging information. I'd say that you just write all of the exceptions and then at the end write `return ('Internal Error', 'danger')` and be done.